server/src/main/java/org/apache/iotdb/db/metadata/mtree/MTreeAboveSG.java

```diff
@@ -0,0 +1,533 @@
+import org.apache.iotdb.db.metadata.path.PartialPath;
+import org.apache.iotdb.db.metadata.template.Template;
+import org.apache.iotdb.db.metadata.utils.MetaFormatUtils;
+import org.apache.iotdb.tsfile.utils.Pair;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static org.apache.iotdb.db.conf.IoTDBConstant.ONE_LEVEL_PATH_WILDCARD;
+
+public class MTreeAboveSG {
+
+  private static final Logger logger = LoggerFactory.getLogger(MTreeAboveSG.class);
+  public static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+
+  private IMNode root;
+
+  public MTreeAboveSG() {
+    this.root = new InternalMNode(null, IoTDBConstant.PATH_ROOT);
+  }
+
+  public void init() {}
+
+  public void clear() {
+    this.root = new InternalMNode(null, IoTDBConstant.PATH_ROOT);
+  }
+
+  /**
+   * Set storage group. Make sure check seriesPath before setting storage group
+   *
+   * @param path path
+   */
+  public IStorageGroupMNode setStorageGroup(PartialPath path) throws MetadataException {
+    String[] nodeNames = path.getNodes();
+    MetaFormatUtils.checkStorageGroup(path.getFullPath());
+    if (nodeNames.length <= 1 || !nodeNames[0].equals(root.getName())) {
+      throw new IllegalPathException(path.getFullPath());
+    }
+    IMNode cur = root;
+    Template upperTemplate = cur.getSchemaTemplate();
+    int i = 1;
+    // e.g., path = root.a.b.sg, create internal nodes for a, b
+    while (i < nodeNames.length - 1) {
+      IMNode temp = cur.getChild(nodeNames[i]);
+      if (temp == null) {
+        if (cur.isUseTemplate() && upperTemplate.hasSchema(nodeNames[i])) {
+          throw new PathAlreadyExistException(
+              cur.getPartialPath().concatNode(nodeNames[i]).getFullPath());
+        }
+        cur.addChild(nodeNames[i], new InternalMNode(cur, nodeNames[i]));
+      } else if (temp.isStorageGroup()) {
+        // before set storage group, check whether the exists or not
+        throw new StorageGroupAlreadySetException(temp.getFullPath());
+      }
+      cur = cur.getChild(nodeNames[i]);
+      upperTemplate = cur.getSchemaTemplate() == null ? upperTemplate : cur.getSchemaTemplate();
+      i++;
+    }
+
+    // synchronize check and add, we need addChild become atomic operation
+    // only write on mtree will be synchronized
+    synchronized (this) {
+      if (cur.hasChild(nodeNames[i])) {
+        // node b has child sg
+        if (cur.getChild(nodeNames[i]).isStorageGroup()) {
+          throw new StorageGroupAlreadySetException(path.getFullPath());
+        } else {
+          throw new StorageGroupAlreadySetException(path.getFullPath(), true);
```

Comment:
What does `hasChild` mean ?

Reply:
If there's already a storage group, named root.a.sg, and users try to set storage group to root.a, then a StorageGroupAlreadySetException with hasChild=true will be thrown. This will forbid MManager from retrying setStorageGroup.

Reply:
Got it